overlay_new.py

```diff
@@ -0,0 +1,197 @@
+from abc import ABC, abstractmethod
+from time import time
+
+from canvas import Canvas, Colour
+from data import Data
+from overlay import Overlay
+from textwrap import wrap
+from typing import Callable, Tuple
+
+class Drawable(ABC):
+    @abstractmethod
+    def draw_base(self, canvas: Canvas):
+        pass
+    @abstractmethod
+    def draw_data(self, canvas: Canvas, data: Data):
+        pass
+
+class DataField(Drawable):
+    """ A simple data field containing a large value below a title in a smaller
+        font. Text is right-aligned. """
+
+    title_size = 0.8
+    data_size = 1.5
+
+    width, data_height = Canvas.get_text_dimensions("100.0", data_size)
+    _, title_height = Canvas.get_text_dimensions("KPH", title_size)
+
+    spacing = 10
+
+    height = title_height + data_height + spacing
+
+    def __init__(self, title: str, value_func: Callable[[Data], str], coordinate: Tuple[int, int]):
+        """ `coordinate` specifies the bottom-left coordinate of the data field. """
+        self.title = title
+        self.value_func = value_func
+        self.title_coord = (coordinate[0] + DataField.width, coordinate[1] - DataField.spacing - DataField.data_height)
+        self.data_coord = (coordinate[0] + DataField.width, coordinate[1])
+
+    def draw_base(self, canvas: Canvas):
+        canvas.draw_text(self.title, self.title_coord, DataField.title_size, Colour.white, "right")
+
+    def draw_data(self, canvas: Canvas, data: Data):
+        canvas.draw_text(self.value_func(data), self.data_coord, DataField.data_size, Colour.white, "right")
+
+class SpeedField(DataField):
+    """ A specialised version of DataField which prefers to display GPS speed,
+        but falls back to reed velocity if GPS is unavailable """
+
+    def __init__(self, coordinate: Tuple[int, int]):
+        value_func = lambda data: "{:.1f}".format(data["gps_speed"] or data["reed_velocity"])
+        super().__init__("", value_func, coordinate)
+
+    def draw_base(self, canvas: Canvas):
+        pass
+
+    def draw_data(self, canvas: Canvas, data: Data):
+        self.title = "GPS KPH" if data["gps_speed"] else "REED KPH"
+        super().draw_base(canvas)
+        super().draw_data(canvas, data)
+
+class CentrePower(Drawable):
+    """ Displays the current and recommended power at the bottom-centre of the
+        screen.
+
+        If the current power is within 5% of recommended, it is coloured green,
+        otherwise red. If no recommended power is available, it is coloured
+        black. """
+
+    power_size = 2.5
+    rec_power_size = 1.5
+    _, rec_power_height = Canvas.get_text_dimensions("500w", rec_power_size)
```

Comment:
Here too~ `"500W` is random?

Reply:
Leave a comment here about the magic string. Variable could help instead of comments?
`random_power_string = "500w"`